datafusion/src/execution/dataframe_impl.rs

```diff
@@ -149,8 +152,19 @@ impl DataFrame for DataFrameImpl {
         Ok(collect(plan).await?)
     }
 
-    // Convert the logical plan represented by this DataFrame into a physical plan and
-    // execute it
+    /// Convert the logical plan represented by this DataFrame into a physical plan and
+    /// execute it, returning a stream over a single partition
+    async fn collect_stream(&self) -> Result<SendableRecordBatchStream> {
+        let state = self.ctx_state.lock().unwrap().clone();
+        let ctx = ExecutionContext::from(Arc::new(Mutex::new(state)));
+        let plan = ctx.optimize(&self.plan)?;
+        let plan = ctx.create_physical_plan(&plan)?;
+        collect_stream(plan).await
+    }
+
+    /// Convert the logical plan represented by this DataFrame into a physical plan and
+    /// execute it, collecting all resulting batches into memory while maintaining
+    /// partitioning
     async fn collect_partitioned(&self) -> Result<Vec<Vec<RecordBatch>>> {
         let state = self.ctx_state.lock().unwrap().clone();
```

Comment:
You could probably rewrite `collect_partitioned` to be in terms of `collect_stream_partitioned`:

```rust
collect(self.collect_stream_partitioned().await?)
``` 
or something like that

Reply:
I've cleaned the code up and removed a fair bit of duplication now.